tests/test_convert_lit_checkpoint.py

```diff
@@ -0,0 +1,59 @@
+from unittest import mock
+from pathlib import Path
+from urllib.request import urlretrieve
+
+import pytest
+import torch
+
+wd = Path(__file__).parent.parent.absolute()
+
+
+def test_convert_lit_checkpoint(tmp_path):
+    from scripts.convert_lit_checkpoint import convert_lit_checkpoint
+
+    chkpt_name = "lit_model_finetuned"
+
+    with pytest.raises(RuntimeError, match="open file failed because of errno 2 on fopen"):
+        convert_lit_checkpoint(checkpoint_name=chkpt_name, checkpoint_dir=tmp_path, model_name="falcon-7b")
+
+    chkpt_filename = tmp_path / "lit_model_finetuned"
+    chkpt_filename.touch()
+    with mock.patch("scripts.convert_lit_checkpoint.lazy_load") as load:
+        convert_lit_checkpoint(checkpoint_name=chkpt_name, checkpoint_dir=tmp_path, model_name="falcon-7b")
+    load.assert_called_with(chkpt_filename)
+
+    assert {p.name for p in tmp_path.glob("*")} == {chkpt_name, "lit_model_finetuned.bin"}
+
+
+@torch.inference_mode()
+def test_against_original_falcon_40b():
+    file_path = wd / "tests" / "original_falcon_40b.py"
+    url = "https://gist.githubusercontent.com/carmocca/feed39b1bc65a29f73c1cecc58a01167/raw/a9a65f2b93716b3c09ec9f354d535ae5953de08f/original_falcon_40b.py"
+    if not file_path.is_file():
+        urlretrieve(url=url, filename=file_path)
+
+    from tests.original_falcon_40b import RWConfig, RWForCausalLM
+    from scripts.convert_lit_checkpoint import copy_weights_falcon as copy_to_theirs
+    from scripts.convert_hf_checkpoint import copy_weights_falcon as copy_to_ours
+
+    config = RWConfig(
+        hidden_size=32, n_head=8, n_head_kv=4, n_layer=2, parallel_attn=True, vocab_size=65024, bias=False
+    )
+
+    their_model = RWForCausalLM(config)
+    their_state_dict = their_model.state_dict()
+
+    ours_state_dict = {}
+    copy_to_ours("40b", ours_state_dict, their_state_dict)
+
+    converted_state_dict = {}
+    copy_to_theirs("40b", converted_state_dict, ours_state_dict)
```

Comment:
Shouldn't this work if we started off our model directly?

```suggestion
    our_model = GPT(our_config)
    ours_state_dict = our_model.state_dict()

    converted_state_dict = {}
    copy_to_theirs("40b", converted_state_dict, ours_state_dict)
```

If not, why?

Reply:
when I do:

```python
    ours_model = GPT(ours_config)
    ours_state_dict = ours_model.state_dict()

    theirs_state_dict = {}
    copy_to_theirs("40b", theirs_state_dict, ours_state_dict)

    theirs_model = RWForCausalLM(theirs_config)
    theirs_model.load_state_dict(theirs_state_dict)

    # test end to end
    x = torch.tensor([[9856, 23, 491, 1536, 304]], dtype=torch.int32)
    theirs_y = theirs_model(x)["logits"]
    ours_y = ours_model(x)
    torch.testing.assert_close(theirs_y, ours_y)
```

I get:

`FAILED tests/test_convert_lit_checkpoint.py::test_against_original_falcon_40b - AssertionError: Tensor-likes are not close!`

with:

```
>       torch.testing.assert_close(converted_y, ours_y)
E       AssertionError: Tensor-likes are not close!
E       
E       Mismatched elements: 325117 / 325120 (100.0%)
E       Greatest absolute difference: 3.43747615814209 at index (0, 0, 58720) (up to 1e-05 allowed)
E       Greatest relative difference: 190444.875 at index (0, 1, 64161) (up to 1.3e-06 allowed)
```

Reply:
I had done something similar to this at first, and this error/failure is why I began writing the assertions to check the length of the state_dicts, the returned param tensors, and that the key names in the converted state_dict matched a ground truth theirs_model.state_dict ... all of those assertions passed, however - the torch.testing.assert_close does not.

Reply:
@carmocca @lantiga this is resolved as of [3e6b412](https://github.com/Lightning-AI/lit-gpt/pull/273/commits/3e6b41208c7385101e84743eedb1b1644821a139).

The fix was setting `assign` to True in `theirs.load_state_dict()`.